examples/ethereum-tracker/src/contract.rs

```diff
@@ -69,69 +75,96 @@ impl Contract for EthereumTrackerContract {
 }
 
 impl EthereumTrackerContract {
-    fn get_endpoints(&self) -> (EthereumClient, String) {
-        let url = self.state.ethereum_endpoint.get().clone();
-        let contract_address = self.state.contract_address.get().clone();
-        let ethereum_client = EthereumClient { url };
-        (ethereum_client, contract_address)
-    }
+    /// Reads the initial event emitted by the Ethereum contract, with the initial account and its
+    /// balance.
+    async fn read_initial(&mut self) {
+        let request = async_graphql::Request::new("query { readInitialEvent }");
 
-    async fn read_initial(&mut self, start_block: u64) {
-        let event_name_expanded = "Initial(address,uint256)";
-        let (ethereum_client, contract_address) = self.get_endpoints();
-        let events = ethereum_client
-            .read_events(
-                &contract_address,
-                event_name_expanded,
-                start_block,
-                start_block + 1,
-            )
-            .await
-            .expect("Read the Initial event");
-        assert_eq!(events.len(), 1);
-        let event = events[0].clone();
-        let EthereumDataType::Address(address) = event.values[0].clone() else {
-            panic!("wrong type for the first entry");
+        let application_id = self.runtime.application_id();
+        let response = self.runtime.query_service(application_id, request);
+
+        let async_graphql::Value::Object(data_object) = response.data else {
```

Comment:
We could avoid all those `async_graphql::Value::Variant` matches by using `response.data.into_json()` instead. The JSON `Value` type has a few more convenient methods like `as_str` etc. (And we'd rarely need `as_object` because e.g. `json_data["readInitialEvent"]["address"]` works, too.)

(Anyway, maybe that's outside the scope of this PR.)